SimPEG/base/pde_simulation.py

```diff
@@ -0,0 +1,529 @@
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mcc{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mcc{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mcc{arg}IDeriv", MccIDeriv_prop)
+
+        def MnIDeriv_prop(self, u, v=None, adjoint=False):
+            """
+            Derivative of `MnPropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mn{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mn{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mn{arg}IDeriv", MnIDeriv_prop)
+
+        def MfIDeriv_prop(self, u, v=None, adjoint=False):
+            """I
+            Derivative of `MfPropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Mf{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Mf{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Mf{arg}IDeriv", MfIDeriv_prop)
+
+        def MeIDeriv_prop(self, u, v=None, adjoint=False):
+            """
+            Derivative of `MePropertyI` with respect to the model.
+            """
+            if getattr(self, f"{arg.lower()}Map") is None:
+                return Zero()
+            if isinstance(u, Zero) or isinstance(v, Zero):
+                return Zero()
+
+            MI_prop = getattr(self, f"Me{arg}I")
+            u = MI_prop @ (MI_prop @ -u)
+            M_prop_deriv = getattr(self, f"Me{arg}Deriv")
+            return M_prop_deriv(u, v, adjoint=adjoint)
+
+        setattr(cls, f"Me{arg}IDeriv", MeIDeriv_prop)
+
+        @property
+        def _clear_on_prop_update(self):
+            items = [
+                f"_Mcc_{arg}",
+                f"_Mn_{arg}",
+                f"_Mf_{arg}",
+                f"_Me_{arg}",
+                f"_MccI_{arg}",
+                f"_MnI_{arg}",
+                f"_MfI_{arg}",
+                f"_MeI_{arg}",
+                f"_Mcc_{arg}_deriv",
+                f"_Mn_{arg}_deriv",
+                f"_Mf_{arg}_deriv",
+                f"_Me_{arg}_deriv",
+            ]
+            return items
+
+        setattr(cls, f"_clear_on_{arg.lower()}_update", _clear_on_prop_update)
+        return cls
+
+    return decorator
+
+
+class BasePDESimulation(BaseSimulation):
```

Comment:
I am wondering if there is a way that would make sense to break this up to keep the name-spaces a bit smaller... e.g. the EM simulations don't need Mcc / Mn (unless we need to solve an initial condition for a grounded source in TDEM, but we might want to give that scenario a bit more thought anyways... it would be nice if we could make use of the DC simulation for this rather than what we are doing right now, which is re-implementing it... but this is a tangent :) 

Anyways, my main question here is: do we want to break this up into `Scalar` and `Vector` PDE simulations? or is that just a level of modularity that isn't helpful.

Reply:
This begs the question for me, in should  *any* of the Mass Matrices be easily accessible to the user in the namespace of the `Simulation`. In my opinion they should start with an `_`, but as a start I was erring on consistency with the current method.

Reply:
Yeah, I was thinking about that too... In the earlier days of simpeg, I would have leaned towards saying they should be top-level accessible for the educational value, but really now, if you want to learn about building up systems of equations, that should be done at the level of discretize, not simpeg. So I would be supportive of moving towards `_` for the mass matrices

Reply:
I could probably tackle that pretty quickly with this PR